### PR TITLE
GH-41034: [C++][FS][Azure] Adjust DeleteDir/DeleteDirContents/GetFileInfoSelector behaviors against Azure for generic filesystem tests

### PR DESCRIPTION
**Pull request by kou**
### Rationale for this change

They are failing:

```text
[  FAILED  ] TestAzureHierarchicalNSGeneric.DeleteDir
[  FAILED  ] TestAzureHierarchicalNSGeneric.DeleteDirContents
[  FAILED  ] TestAzureHierarchicalNSGeneric.GetFileInfoSelector
```

### What changes are included in this PR?

* `DeleteDir()`: Check not a directory case
* `DeleteDirContents()`: Check not a directory case
* `GetFileInfoSelector()`:
  * Add not a directory check for input
  * Add support for returning metadata for directory 

### Are these changes tested?

Yes.

### Are there any user-facing changes?

Yes.
* GitHub Issue: #41034

### CODE REVIEW

**Comment by kou** [comment]
@felipecrv It seems that the current `CopyFile` implementation doesn't work with Azure hierarchical namespace support. I'll split it to a separated issue.

File: cpp/src/arrow/filesystem/azurefs.cc [hunk 1]
```diff
@@ -1642,11 +1642,27 @@ class AzureFileSystem::Impl {
...
+      } else if (info.type() != FileType::Directory) {
+        return NotADir(base_location);
+      }
+      options.Prefix = prefix.path;
     }
     options.PageSizeHint = page_size_hint;
     options.Include = Blobs::Models::ListBlobsIncludeFlags::Metadata;
 
+    auto adlfs_client = GetFileSystemClient(base_location.container);
+    ARROW_ASSIGN_OR_RAISE(auto hns_support, HierarchicalNamespaceSupport(adlfs_client));
```
Comment (felipecrv) [inline cpp/src/arrow/filesystem/azurefs.cc:1664]: This fix is very confusing to me. This function is `GetFileInfoWithSelectorFromContainer`, it's meant to be called when HNS support has been detected as disabled.

The right fix (IMO) is to create `GetFileInfoWithSelectorFromFileSystem(adlfs_client, ...)` and then in `GetFileInfoWithSelector` dispatch to the different implementations according to the HNS support detection.

File: cpp/src/arrow/filesystem/azurefs.cc [hunk 2]
```diff
@@ -2168,13 +2203,6 @@ class AzureFileSystem::Impl {
       // All the others either succeed or throw an exception.
       DCHECK(response.Value.Deleted);
     } catch (const Storage::StorageException& exception) {
-      if (exception.ErrorCode == "FilesystemNotFound" ||
-          exception.ErrorCode == "PathNotFound") {
-        if (require_dir_to_exist) {
-          return PathNotFound(location);
-        }
-        return Status::OK();
-      }
```
Comment (felipecrv) [inline cpp/src/arrow/filesystem/azurefs.cc:2177]: The fact that you check existence above doesn't mean these errors can't happen down here.

To use terms that @pitrou once shared with me [1], you're moving from EAFP [2] "easier to ask for permission" to LBYL [3] "look before you leap".

From [3]:
> ### LBYL
> Look before you leap. This coding style explicitly tests for pre-conditions before making calls or lookups. This style contrasts with the [EAFP](https://docs.python.org/3.5/glossary.html#term-eafp) approach and is characterized by the presence of many [if](https://docs.python.org/3.5/reference/compound_stmts.html#if) statements.
>
> In a multi-threaded environment, **the LBYL approach can risk introducing a race condition** between “the looking” and “the leaping”. For example, the code, if key in mapping: return mapping[key] can fail if another thread removes key from mapping after the test, but before the lookup. This issue can be solved with locks or by using the EAFP approach.


[1] https://devblogs.microsoft.com/python/idiomatic-python-eafp-versus-lbyl/
[2] https://docs.python.org/3.5/glossary.html#term-eafp
[3] https://docs.python.org/3.5/glossary.html#term-lbyl

Comment (felipecrv) [reply cpp/src/arrow/filesystem/azurefs.cc:2177]: Given the TZ differences, I will give myself a chance of fixing these bugs to see if I can come up with something that doesn't introduce pessimistic round trips to the backend.

File: cpp/src/arrow/filesystem/azurefs.cc (hunk 1 above)
Comment (kou) [reply cpp/src/arrow/filesystem/azurefs.cc:1664]: I can understand what you say but the current implementation doesn't:

`GetFileInfoWithSelector()` isn't dispatched based on HNS support:

https://github.com/apache/arrow/blob/cd607d00b9e3c5f2ac2a2e373ba7571fd75fc725/cpp/src/arrow/filesystem/azurefs.cc#L2859-L2866

https://github.com/apache/arrow/blob/cd607d00b9e3c5f2ac2a2e373ba7571fd75fc725/cpp/src/arrow/filesystem/azurefs.cc#L1740-L1779

The `FromContainer` suffix doesn't mean only for no HNS support. We use `OnContainer`/`OnFileSystem` suffix for it. `FromContainer` just means that it's for one container and it doesn't work with multiple containers.

----

Should we create `GetFileInfoWithSelectorFromContainerOn{Container,FileSystem}()`? But their names are too long...

File: cpp/src/arrow/filesystem/azurefs.cc (hunk 2 above)
Comment (kou) [reply cpp/src/arrow/filesystem/azurefs.cc:2177]: Can we block changing the target location from other connections by using `lease_id` for `GetFileInfo()` too? 

```diff
diff --git a/cpp/src/arrow/filesystem/azurefs.cc b/cpp/src/arrow/filesystem/azurefs.cc
index df6db3d959..67232b3156 100644
--- a/cpp/src/arrow/filesystem/azurefs.cc
+++ b/cpp/src/arrow/filesystem/azurefs.cc
@@ -2181,7 +2181,7 @@ class AzureFileSystem::Impl {
                                Azure::Nullable<std::string> lease_id = {}) {
     DCHECK(!location.container.empty());
     DCHECK(!location.path.empty());
-    ARROW_ASSIGN_OR_RAISE(auto file_info, GetFileInfo(adlfs_client, location));
+    ARROW_ASSIGN_OR_RAISE(auto file_info, GetFileInfo(adlfs_client, location, lease_id));
     if (file_info.type() == FileType::NotFound) {
       if (require_dir_to_exist) {
         return PathNotFound(location);
```

File: cpp/src/arrow/filesystem/azurefs.cc (hunk 1 above)
Comment (felipecrv) [reply cpp/src/arrow/filesystem/azurefs.cc:1664]: > I can understand what you say but the current implementation doesn't:
>
> GetFileInfoWithSelector() isn't dispatched based on HNS support:

But what I meant was changing/fixing `GetFileInfoWithSelector` to do that.

> Should we create GetFileInfoWithSelectorFromContainerOn{Container,FileSystem}()? But their names are too long...

ADLFS calls "filesystem" what "blobs" calls "container", so you replace the last word of the name:

```cpp
  Status GetFileInfoWithSelectorFromContainer(const Blobs::BlobContainerClient& container_client, ...);
  Status GetFileInfoWithSelecitonFromFilesystem(const DataLake::DataLakeFileSystemClient& adlfs_client, ...);
 ```

File: cpp/src/arrow/filesystem/azurefs.cc (hunk 2 above)
Comment (felipecrv) [reply cpp/src/arrow/filesystem/azurefs.cc]: > But DeleteRecursive()/DeleteEmpty() can delete both of file and directory. See also: https://learn.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/delete?view=rest-storageservices-datalakestoragegen2-2019-12-12

This scenario sounds familiar. 🤔  Does that happen even if you append a trailing slash to the path?

Comment (kou) [reply cpp/src/arrow/filesystem/azurefs.cc]: If we add a trailing slash, API returns an error...:

```text
'fs->DeleteDir("AB/CD")' failed with IOError: Failed to delete a directory: AB/CD: https://clearcodearrow.blob.core.windows.net/aeouspyxzsz16iliz71ya6yt5qrur6b9/AB/CD/ Azure Error: [InvalidUri] 400 The request URI is invalid.
```

File: cpp/src/arrow/filesystem/azurefs.cc (hunk 1 above)
Comment (kou) [reply cpp/src/arrow/filesystem/azurefs.cc:1664]: > ADLFS calls "filesystem" what "blobs" calls "container"

Wow! I didn't notice it!

Comment (kou) [reply cpp/src/arrow/filesystem/azurefs.cc]: Implemented.

**Comment by kou** [comment]
@felipecrv Do you want to review this before we merge this? I want to add this to 16.0.0.

**Review by felipecrv — APPROVED** [review]
Thank you! This looks much more straightforward now.

**Comment by kou** [comment]
Thanks! I'll merge this.

**Pull request merged**